### PR TITLE
setting up framework for browser-side testing with Chai

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

File: test.array.js [hunk 1]
```diff
@@ -0,0 +1,25 @@
+var assert = require('assert');
```
Comment (rjz) [inline test.array.js]: This include assumes that a node-style module system is available and will require some extra tooling to work in the browser, however we can work around it here for now. `chai` (included below) is an assertion library that is added to the global scope by the runner below...

File: test.array.js [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+    before(function() {
+      fakeArray[0] = 1;
+      fakeArray[1] = 2;
+      fakeArray[2] = 3;
+      fakeArray.length = 3;
+    });
+
+    it('should return the final element', function() {
+      assert.equal(fakeArray.pop(), 3);
```
Comment (rjz) [inline test.array.js]: ...and with a few caveats in the runner below, we can rewrite this assertion using chai's `expect()` syntax as:

```
chai.expect(fakeArray.pop()).to.eq(3);
```

File: test.array.js [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    });
+  });
+});
+
+describe('Set Origin Button', function(){
+  it('should bring up modal', function(){
+    expect.
```
Comment (rjz) [inline test.array.js]: :^)

File: views/home.html [hunk 4]
```diff
@@ -44,6 +45,23 @@
        </div>
      </div>
 
+     <div id="mocha"></div>
+
+    <script src="https://cdn.rawgit.com/jquery/jquery/2.1.4/dist/jquery.min.js"></script>
+    <script src="https://cdn.rawgit.com/Automattic/expect.js/0.3.1/index.js"></script>
```
Comment (rjz) [inline views/home.html]: expect.js is a (light) analogue to mocha/chai—if we want to use the latter this should be safe to remove.

**Comment by ghost** [comment]
Awesome, thanks RJ. The initial array test ran without a hitch so it looks like the mocha framework and chai assertion library are all working properly. I'm still pretty new to writing automated tests and everything we've done in the class so far has been through the command line so I was wondering if you had a few suggestions for the best strategy on writing tests for the browser. 

For example, I have this test case in the list:

Notes Modal
- clicking notes should change status/modify CSS
- clear Notes button should remove notes with status of 'superDone' 

..but given that we're using Google's API, Backbone, Bootstrap and HTML/CSS to coordinate these events, I'm not too sure on where to even begin on writing a test for that or how conceptualize the click events when one part is in the HTML while the other is in Backbone.

Thanks again!

**Comment by rjz** [comment]
Glad to hear it's all playing nicely together!

> ..but given that we're using Google's API, Backbone, Bootstrap and HTML/CSS to coordinate these events, I'm not too sure on where to even begin on writing a test for that or how conceptualize the click events when one part is in the HTML while the other is in Backbone.

#### Background

This is a huge problem in testing ("what level should we test at?") and there are a few options:
- **unit tests**, which test a single "unit" of software--a function, class, or module. In unit testing the goal is to isolate each unit from its dependencies and assert that it behaves in a certain way. These tests tend to be cheap to write and fast to run, but more expensive to maintain: any time any unit in the codebase changes, that behavior must be adjusted in both all related units and their tests.
- **functional tests**, which test several elements of the same piece of software. A RESTful API would be a good candidate for a functional test; given a certain HTTP request (or sequence of HTTP requests), such tests will verify the HTTP response and any changes to the application state. These are more expensive to write, middling-slow to run, but less expensive to maintain: since they test the relatively static "public interface" of an application, its internals may be easily changed without affecting its functional tests.
- **acceptance tests**, which test the behavior of a completed system across its constituent pieces. At the simple end this might be the client and server of a web application; in more complex applications this might involve testing across multiple client-side pieces and many backend services. These can be _very_ expensive to write and run, but useful nevertheless for verifying certain critical, high-level features.

In general, the number of tests at each level follows the cost. We'll write many unit tests, some functional tests, and a limited set of acceptance tests. It's not a hard and fast rule, though: our codebase is very heavy on functional tests and relatively light on unit tests. We still write them to guide each unit's development--but rather than paying their upkeep cost, we are quick to throw them out in favor of higher-level functional tests even before a features is merged.

#### Testing the Modal

With all that in mind, let's look at the modal example:

> Notes Modal
> - clicking notes should change status/modify CSS
> - clear Notes button should remove notes with status of 'superDone'

These are both phrased as fairly high-level tests, but we should be able to implement them mostly as unit tests. If I'm mapping terminology correctly, the "unit" involved here won't actually be the modal, but the `NotesCollectionView` it wraps.

Stubbing out the tests, we could write a BDD-style spec:

```
describe('NotesCollectionView', function () {
  describe('when a note is clicked', function () {
    it('updates CSS');
    it('resets input);
    it('adds new note to the collection');
  });

  describe('when the "clear" button is clicked', function () {
    it('removes `superDone` notes');
  });
});
```

If we run the tests now, we'll see several "pending" tests (i.e., the assertions have not been defined yet).

#### Stubbing dependencies

The next part of the puzzle is to set up the tests such that we have an isolated `NotesCollectionView`. It actually starts life in fairly good shape, though there are a few external dependencies:
- the DOM
- Backbone behaviors

Both the DOM and `Backbone` are global elements. This makes them difficult to isolate completely, so we'll only focus on the pieces we're concerned about. For the DOM, that means the element the view is attached to; for `Backbone` that means the `sync` adapter. Since we get to specify what `$el` we supply to the view (this is sometimes called an "injected" dependency), this part is straightforward. We'll just create our own DOM fragment and supply it to the view:

```
describe('when a note is clicked', function () {
  var collection = $el = view = null;

  // set up the view
  beforeEach(function () {
    collection = new NotesCollection();
    $el = $('<div />');
    view = new NotesCollectionView({
      collection: collection,
      $el: el
    });
    view.render();
  });

  // tear down the view
  afterEach(function () {
    $el.remove();
  });

  it('updates CSS');
  it('resets input');
  it('adds new note to the collection');
});
```

Working around Backbone is a little bit trickier, since we're actually interested in how the view interacts with its collection. Fortunately, we're still supplying the collection argument, so we can once again take advantage of an injected dependency. The goal will be to replace a method whose behavior we don't want to deal with in this test (e.g. `collection.create`) with a "stubbed" version that's easier to test. Rather than triggering a server update, maybe we only care that the particular method was called.

```
// set up the view
beforeEach(function () {
  collection = new NotesCollection();
  collection.createCallCount = 0;
  collection.create = function () {
    this.createCallCount++;
  };

  // ...
});

// ...
```

Ideally we won't _change_ the behavior (return type, side effects) of the `collection.create` method, but in the spirit of demonstration we've simply stubbed in an incrementing counter. We also don't generally need to do this ourselves: [sinon.js](http://sinonjs.org/) is a widely-used library with a ton of utilities supporting this sort of operation.

#### Implementing the test

Anyway, we're now ready to actually write a test. We'll begin with one more setup step that "clicks" the input:

```
// ...

// add an input and click the button
beforeEach(function () {
  view.$('#notesInput').val('Some test input');
  view.$('#tskBtn').click();
});

// ...
```

Now we can go about actually implementing the "`it`" assertions. Since the setup is in a `beforeEach`, our resetting test (for instance) will actually expect to see an empty input:

```
it('resets input', function () {
  expect(view.$('#notesInput').val()).to.eq('');
});

it('adds new note to the collection', function () {
  expect(view.collection.createCallCount).to.eq(1);
});
```

And if everything went well, two of the pending tests should now run green!

Hopefully this covers the immediate need. One big item to keep an eye on is the "testability" of various units within the code: as long as dependencies are injected through function arguments or a constructor, it's fairly easy to isolate the unit in the tests. As soon as global elements are pulled in, though (looking at you, google maps API) the same isolation requires quite a bit more creativity.

Node, ES6, and most other languages provides module systems to help encourage modularity, but for now these must be bolted on to browser applications. It might be worth looking into using something like [browserify](http://browserify.org/) here if the project continues past the end of class (please reach out before you set it up—we can save some time on that!), but for the time being it's more of a pattern to be aware of than anything else.

**Pull request closed without merging**